File: HOT100/8.二叉树/114.二叉树展开为链表.py
#
# @lc app=leetcode.cn id=114 lang=python3
#
# [114] 二叉树展开为链表
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        def dfs(n):
            if not n: return
            if n.left:
                tmp = n.right
                n.right = n.left
                n.left = None
                u = n
                while u.right:
                    u = u.right
                u.right = tmp
            dfs(n.right)
        return dfs(root)
# @lc code=end

File: HOT100/8.二叉树/114.二叉树展开为链表.cpp
/*
 * @lc app=leetcode.cn id=114 lang=cpp
 *
 * [114] 二叉树展开为链表
 */
#include "tools.h"
struct TreeNode
{
    int val;
    TreeNode *left, *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    void flatten(TreeNode *root)
    {
        TreeNode *n = root;
        while (n)
        {
            if (n->left)
            {
                TreeNode *tmpr = n->right;
                TreeNode *cur = n->left;
                TreeNode *tmpl = cur;
                while (cur->right) cur = cur->right;
                n->left = NULL;
                n->right = tmpl;
                cur->right = tmpr;
            }
            n = n->right;
        }
    }
};
// @lc code=end

class Solution
{
public:
    void flatten(TreeNode *root)
    {

        auto dfs = [&](auto &&self, TreeNode *n)
        {
            if (!n)
                return;
            if (n->left)
            {
                TreeNode *tmpr = n->right;
                TreeNode *cur = n->left;
                n->right = cur;
                while (cur->right)
                    cur = cur->right;
                cur->right = tmpr;
                n->left = NULL;
            }
            self(self, n->right);
        };
        dfs(dfs, root);
    }
};